src/storage/lock_manager/deadlock.rs

```diff
@@ -23,83 +23,116 @@ use std::fmt::{self, Display, Formatter};
+    pub fn new(ttl: time::Duration) -> Self {
+        Self {
+            wait_for_map: HashMap::default(),
+            ttl,
+            last_active_expire: time::Instant::now_coarse(),
+            now: time::Instant::now_coarse(),
+        }
+    }
+
+    /// Returns the key hash which causes deadlock.
     pub fn detect(&mut self, txn_ts: u64, lock_ts: u64, lock_hash: u64) -> Option<u64> {
         let _timer = DETECTOR_HISTOGRAM_VEC.detect.start_coarse_timer();
         TASK_COUNTER_VEC.detect.inc();
 
-        // If txn_ts is waiting for lock_ts, it won't cause deadlock.
+        self.now = time::Instant::now_coarse();
+        self.active_expire();
+
+        // If `txn_ts` is waiting for `lock_ts`, it won't cause deadlock.
         if self.register_if_existed(txn_ts, lock_ts, lock_hash) {
             return None;
         }
 
-        // Memorize the searching path.
-        let mut visited: HashSet<u64> = HashSet::default();
-        if let Some(deadlock_key_hash) = self.do_detect(&mut visited, txn_ts, lock_ts) {
+        if let Some(deadlock_key_hash) = self.do_detect(txn_ts, lock_ts) {
             return Some(deadlock_key_hash);
         }
         self.register(txn_ts, lock_ts, lock_hash);
         None
     }
 
-    /// Checks if there is a path from `wait_for_ts` to `txn_ts`.
-    fn do_detect(&self, visited: &mut HashSet<u64>, txn_ts: u64, wait_for_ts: u64) -> Option<u64> {
-        visited.insert(wait_for_ts);
-        if let Some(wait_for) = self.wait_for_map.get(&wait_for_ts) {
-            for (lock_ts, lock_hashes) in wait_for {
-                if *lock_ts == txn_ts {
-                    return Some(lock_hashes[0]);
-                }
-                if visited.contains(lock_ts) {
-                    continue;
-                }
-                if let Some(deadlock_key_hash) = self.do_detect(visited, txn_ts, *lock_ts) {
-                    return Some(deadlock_key_hash);
+    /// Checks if there is a edge from `wait_for_ts` to `txn_ts`.
+    fn do_detect(&mut self, txn_ts: u64, wait_for_ts: u64) -> Option<u64> {
+        let now = self.now;
+        let ttl = self.ttl;
+
+        // Memorize the searched vertexes.
+        let mut visited: HashSet<u64> = HashSet::default();
+        let mut stack = vec![wait_for_ts];
+        while let Some(wait_for_ts) = stack.pop() {
+            if let Some(wait_for) = self.wait_for_map.get_mut(&wait_for_ts) {
+                // Remove expired edges.
+                wait_for.retain(|_, (_, time)| now.duration_since(*time) < ttl);
+                if wait_for.is_empty() {
+                    self.wait_for_map.remove(&wait_for_ts);
+                } else {
+                    for (lock_ts, (lock_hashes, _)) in wait_for {
+                        if *lock_ts == txn_ts {
+                            return Some(lock_hashes[0]);
+                        }
+                        if !visited.contains(lock_ts) {
+                            stack.push(*lock_ts);
```

Comment:
How about the following map:
1->2->3
1->4
seems in this map, the note 4 will be pushed into the stack twice.

Reply:
Oh,  it seems `visited.insert` should be invoked when pushing to stack?

Reply:
Yes.

Reply:
Hi~ I'm fresher of rust and tikv and I'm trying to study tikv distribution transaction. I've tried hard but still can't understand this function clearly since it has very high O complexity. Is there any way to simplify this code so that it's more friendly to fresher like me?

Reply:
Maybe less conditions`{}`in conditions`{}` by:
1. return or continue as soon as possible.
2. extract more function
      -- by <<Refactoring, improving the design of existing code>>

Does the following example easier for review? which is instead for L128~146.

```
 while let Some(wait_for_ts) = stack.pop() {
            let wait_for = if let Some(wait_for) = self.wait_for_map.get_mut(&wait_for_ts) {
                wait_for
            } else {
                continue;
            };
           
            // Remove expired edges.
            wait_for.retain(|_, locks| !locks.is_expired(now, ttl));
            if wait_for.is_empty() {
                self.wait_for_map.remove(&wait_for_ts);
                continue;
            } 
            // BFS edges
            for (lock_ts, locks) in wait_for {
                if *lock_ts == txn_ts {
                    return Some(locks.hashes[0]);
                }
                if !pushed.contains(lock_ts) {
                    stack.push(*lock_ts);
                    pushed.insert(*lock_ts);
                }
            } 
        }
```

Reply:
The detect table is a DAG, each detect request will add an edge to it. Deadlock means there is a cycle in the table.
As the comment says, the function is to check if there is an edge from `wait_for_ts` to `txn_ts`. If there is, it will cause deadlock. It's a simple iterative DFS.